core/src/main/java/com/linecorp/armeria/server/healthcheck/ScheduledHealthChecker.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the Licenses
+ */
+
+package com.linecorp.armeria.server.healthcheck;
+
+import java.time.Duration;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Futures;
+
+import com.linecorp.armeria.common.util.AbstractListenable;
+
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * A {@link ListenableHealthChecker} whose state can be set by the result of supplied CompletionStage.
+ * This class can be shared by multiple {@link HealthCheckService}.
+ */
+final class ScheduledHealthChecker extends AbstractListenable<HealthChecker>
+        implements ListenableHealthChecker {
+
+    private final Supplier<? extends CompletionStage<HealthCheckStatus>> healthChecker;
+    private final Duration fallbackTtl;
+    private final EventExecutor eventExecutor;
+    private final Consumer<HealthChecker> onHealthCheckerUpdate;
+    private final AtomicBoolean isHealthy;
+    private final AtomicInteger requestCount;
+    private final AtomicReference<ScheduledHealthCheckerImpl> impl;
+
+    ScheduledHealthChecker(Supplier<? extends CompletionStage<HealthCheckStatus>> healthChecker,
+                           Duration facllbackTtl, EventExecutor eventExecutor) {
+        this.healthChecker = healthChecker;
+        this.fallbackTtl = facllbackTtl;
+        this.eventExecutor = eventExecutor;
+
+        isHealthy = new AtomicBoolean();
+        requestCount = new AtomicInteger();
+        impl = new AtomicReference<>();
+        onHealthCheckerUpdate = latestValue -> {
+            isHealthy.set(latestValue.isHealthy());
+            notifyListeners(latestValue);
+        };
+    }
+
+    @Override
+    public boolean isHealthy() {
+        return isHealthy.get();
+    }
+
+    void startHealthChecker() {
+        if (requestCount.getAndIncrement() != 0) {
+            return;
+        }
+        final ScheduledHealthCheckerImpl newlyScheduled =
+                new ScheduledHealthCheckerImpl(healthChecker, fallbackTtl, eventExecutor);
```

Comment:
Should we early instantiate `ScheduledHealthCheckerImpl` when `ScheduledHealthChecker` is created.
`ScheduledHealthCheckerImpl` is a small object. We may not take an advantage of this lazy initialization.

Reply:
We have some situation like stop->restart server, so at that case we need to recreate the new impl instance(I think it's easier that throwing away the stopped impl than reset the state of impl.) So even I move the instance creation to constructor, I still need to keep it recreate here.

Reply:
Question: Could we reuse `ScheduledHealthCheckerImpl` if only it has two states such as `STOPPED`, `STARTED`.

Reply:
Maybe yes, it means we can delete ScheduledHealthCheckerImpl and put the logic into ScheduledHealthChecker.
But I need to recheck the logic, because we need to handle A start -> A stop but not complete yet -> B start case as comment in code(e.g. cancel in progress CompletableFuture and make new one run). Let me think if the code will become cleaner or not 🤔

Reply:
My current concern is for example we have two tasks(one is trigger the supplier, one is the user's CompletableFuture task). If we use the same instance of `ScheduledHealthCheckerImpl`, how to prevent the stopped task triggerring next task. Currently I make the previous instance stop, so even schedule from stopped `ScheduledHealthCheckerImpl`, it will stop at next `runHealthCheck`.
Or we need to make a synchronized block at `start`/`stop`/`callback in CompletableFuture#handle`to avoid duplicate scheduling of `runHealthCheck`.